In [1]:
 import pandas as pd

In [2]:
from pandas import DataFrame,Series

In [3]:
import numpy as np

Importing car_sales data set

In [4]:
df=pd.read_csv("Car_sales_details_P3.csv")

In [5]:
df.head()

Sales_ID                          name  year  selling_price  km_driven  \
0         1        Maruti Swift Dzire VDI  2014         450000     145500   
1         2  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000   
2         3      Honda City 2017-2020 EXi  2006         158000     140000   
3         4     Hyundai i20 Sportz Diesel  2010         225000     127000   
4         5        Maruti Swift VXI BSIII  2007         130000     120000   

      State or Province           City    fuel seller_type transmission  \
0  District of Columbia     Washington  Diesel  Individual       Manual   
1              New York  New York City  Diesel  Individual       Manual   
2              Illinois        Chicago  Petrol  Individual       Manual   
3              Illinois        Chicago  Diesel  Individual       Manual   
4              New York  New York City  Petrol  Individual       Manual   

          owner     mileage   engine   max_power  seats  
0   First Owner   23.4 kmpl  1248 CC      74 bhp      5  
1  Second Owner  21.14 kmpl  1498 CC  103.52 bhp      5  
2   Third Owner   17.7 kmpl  1497 CC      78 bhp      5  
3   First Owner   23.0 kmpl  1396 CC      90 bhp      5  
4   First Owner   16.1 kmpl  1298 CC    88.2 bhp      5

In [6]:
df

Sales_ID                          name  year  selling_price  km_driven  \
0            1        Maruti Swift Dzire VDI  2014         450000     145500   
1            2  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000   
2            3      Honda City 2017-2020 EXi  2006         158000     140000   
3            4     Hyundai i20 Sportz Diesel  2010         225000     127000   
4            5        Maruti Swift VXI BSIII  2007         130000     120000   
...        ...                           ...   ...            ...        ...   
7901      8124             Hyundai i20 Magna  2013         320000     110000   
7902      8125         Hyundai Verna CRDi SX  2007         135000     119000   
7903      8126        Maruti Swift Dzire ZDi  2009         382000     120000   
7904      8127               Tata Indigo CR4  2013         290000      25000   
7905      8128               Tata Indigo CR4  2013         290000      25000   

         State or Province           City    fuel seller_type transmission  \
0     District of Columbia     Washington  Diesel  Individual       Manual   
1                 New York  New York City  Diesel  Individual       Manual   
2                 Illinois        Chicago  Petrol  Individual       Manual   
3                 Illinois        Chicago  Diesel  Individual       Manual   
4                 New York  New York City  Petrol  Individual       Manual   
...                    ...            ...     ...         ...          ...   
7901                 Texas          Plano  Petrol  Individual       Manual   
7902                 Texas    Port Arthur  Diesel  Individual       Manual   
7903                 Texas    Port Arthur  Diesel  Individual       Manual   
7904         Massachusetts      Lunenburg  Diesel  Individual       Manual   
7905                 Maine        Augusta  Diesel  Individual       Manual   

                     owner     mileage   engine   max_power  seats  
0              First Owner   23.4 kmpl  1248 CC      74 bhp      5  
1             Second Owner  21.14 kmpl  1498 CC  103.52 bhp      5  
2              Third Owner   17.7 kmpl  1497 CC      78 bhp      5  
3              First Owner   23.0 kmpl  1396 CC      90 bhp      5  
4              First Owner   16.1 kmpl  1298 CC    88.2 bhp      5  
...                    ...         ...      ...         ...    ...  
7901           First Owner   18.5 kmpl  1197 CC   82.85 bhp      5  
7902  Fourth & Above Owner   16.8 kmpl  1493 CC     110 bhp      5  
7903           First Owner   19.3 kmpl  1248 CC    73.9 bhp      5  
7904           First Owner  23.57 kmpl  1396 CC      70 bhp      5  
7905           First Owner  23.57 kmpl  1396 CC      70 bhp      5  

[7906 rows x 15 columns]

Checking whether there exists any null values for any feature

In [7]:
df.isnull().sum()

Sales_ID             0
name                 0
year                 0
selling_price        0
km_driven            0
State or Province    0
City                 0
fuel                 0
seller_type          0
transmission         0
owner                0
mileage              0
engine               0
max_power            0
seats                0
dtype: int64

In [8]:
df1=pd.read_csv("sales_Status_p3.csv")

In [9]:
df1

Sales_ID sold
0            1    Y
1            2    Y
2            3    Y
3            4    Y
4            5    Y
...        ...  ...
7901      8124    N
7902      8125    N
7903      8126    N
7904      8127    N
7905      8128    N

[7906 rows x 2 columns]

In [10]:
df2=pd.merge(df, df1, on='Sales_ID', how='inner')


In [11]:
df.head()

Sales_ID                          name  year  selling_price  km_driven  \
0         1        Maruti Swift Dzire VDI  2014         450000     145500   
1         2  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000   
2         3      Honda City 2017-2020 EXi  2006         158000     140000   
3         4     Hyundai i20 Sportz Diesel  2010         225000     127000   
4         5        Maruti Swift VXI BSIII  2007         130000     120000   

      State or Province           City    fuel seller_type transmission  \
0  District of Columbia     Washington  Diesel  Individual       Manual   
1              New York  New York City  Diesel  Individual       Manual   
2              Illinois        Chicago  Petrol  Individual       Manual   
3              Illinois        Chicago  Diesel  Individual       Manual   
4              New York  New York City  Petrol  Individual       Manual   

          owner     mileage   engine   max_power  seats  
0   First Owner   23.4 kmpl  1248 CC      74 bhp      5  
1  Second Owner  21.14 kmpl  1498 CC  103.52 bhp      5  
2   Third Owner   17.7 kmpl  1497 CC      78 bhp      5  
3   First Owner   23.0 kmpl  1396 CC      90 bhp      5  
4   First Owner   16.1 kmpl  1298 CC    88.2 bhp      5

In [12]:
df2

Sales_ID                          name  year  selling_price  km_driven  \
0            1        Maruti Swift Dzire VDI  2014         450000     145500   
1            2  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000   
2            3      Honda City 2017-2020 EXi  2006         158000     140000   
3            4     Hyundai i20 Sportz Diesel  2010         225000     127000   
4            5        Maruti Swift VXI BSIII  2007         130000     120000   
...        ...                           ...   ...            ...        ...   
7901      8124             Hyundai i20 Magna  2013         320000     110000   
7902      8125         Hyundai Verna CRDi SX  2007         135000     119000   
7903      8126        Maruti Swift Dzire ZDi  2009         382000     120000   
7904      8127               Tata Indigo CR4  2013         290000      25000   
7905      8128               Tata Indigo CR4  2013         290000      25000   

         State or Province           City    fuel seller_type transmission  \
0     District of Columbia     Washington  Diesel  Individual       Manual   
1                 New York  New York City  Diesel  Individual       Manual   
2                 Illinois        Chicago  Petrol  Individual       Manual   
3                 Illinois        Chicago  Diesel  Individual       Manual   
4                 New York  New York City  Petrol  Individual       Manual   
...                    ...            ...     ...         ...          ...   
7901                 Texas          Plano  Petrol  Individual       Manual   
7902                 Texas    Port Arthur  Diesel  Individual       Manual   
7903                 Texas    Port Arthur  Diesel  Individual       Manual   
7904         Massachusetts      Lunenburg  Diesel  Individual       Manual   
7905                 Maine        Augusta  Diesel  Individual       Manual   

                     owner     mileage   engine   max_power  seats sold  
0              First Owner   23.4 kmpl  1248 CC      74 bhp      5    Y  
1             Second Owner  21.14 kmpl  1498 CC  103.52 bhp      5    Y  
2              Third Owner   17.7 kmpl  1497 CC      78 bhp      5    Y  
3              First Owner   23.0 kmpl  1396 CC      90 bhp      5    Y  
4              First Owner   16.1 kmpl  1298 CC    88.2 bhp      5    Y  
...                    ...         ...      ...         ...    ...  ...  
7901           First Owner   18.5 kmpl  1197 CC   82.85 bhp      5    N  
7902  Fourth & Above Owner   16.8 kmpl  1493 CC     110 bhp      5    N  
7903           First Owner   19.3 kmpl  1248 CC    73.9 bhp      5    N  
7904           First Owner  23.57 kmpl  1396 CC      70 bhp      5    N  
7905           First Owner  23.57 kmpl  1396 CC      70 bhp      5    N  

[7906 rows x 16 columns]

In [13]:
df3=pd.read_csv("State_region_mapping_p3.csv")

In [14]:
df3

Region     State or Province
0   Central              Illinois
1   Central               Indiana
2   Central                  Iowa
3   Central                Kansas
4   Central              Michigan
5   Central             Minnesota
6   Central              Missouri
7   Central              Nebraska
8   Central          North Dakota
9   Central              Oklahoma
10  Central          South Dakota
11  Central                 Texas
12  Central             Wisconsin
13     East           Connecticut
14     East              Delaware
15     East  District of Columbia
16     East                 Maine
17     East              Maryland
18     East         Massachusetts
19     East         New Hampshire
20     East            New Jersey
21     East              New York
22     East                  Ohio
23     East          Pennsylvania
24     East          Rhode Island
25     East               Vermont
26     East         West Virginia
27    South               Alabama
28    South              Arkansas
29    South               Florida
30    South               Georgia
31    South              Kentucky
32    South             Louisiana
33    South           Mississippi
34    South        North Carolina
35    South        South Carolina
36    South             Tennessee
37    South              Virginia
38     West               Arizona
39     West            California
40     West              Colorado
41     West                 Idaho
42     West               Montana
43     West                Nevada
44     West            New Mexico
45     West                Oregon
46     West                  Utah
47     West            Washington
48     West               Wyoming

In [15]:
pd.merge(df2, df3, on='State or Province', how='inner')


Sales_ID                                       name  year  \
0            1                     Maruti Swift Dzire VDI  2014   
1           38           Hyundai Verna VTVT 1.6 SX Option  2019   
2           66                 Ford Figo 1.2P Titanium MT  2017   
3           68  Volkswagen Ameo 1.5 TDI Highline 16 Alloy  2016   
4           69                       Honda City i DTEC VX  2015   
...        ...                                        ...   ...   
7901      4547                       Maruti Ciaz VDI SHVS  2016   
7902      4548                   Maruti Vitara Brezza ZDi  2017   
7903      4549                        Maruti Alto 800 LXI  2016   
7904      6888                    Hyundai Grand i10 Magna  2016   
7905      6889    Maruti Vitara Brezza ZDi Plus Dual Tone  2018   

      selling_price  km_driven     State or Province        City    fuel  \
0            450000     145500  District of Columbia  Washington  Diesel   
1           1149000       5000  District of Columbia  Washington  Petrol   
2            600000      25000  District of Columbia  Washington  Petrol   
3            540000      40000  District of Columbia  Washington  Diesel   
4            630000     135000  District of Columbia  Washington  Diesel   
...             ...        ...                   ...         ...     ...   
7901         620000      90000          South Dakota   Brookings  Diesel   
7902         750000      90000          South Dakota   Brookings  Diesel   
7903         270000      35000          South Dakota   Brookings  Petrol   
7904         470000      10000          South Dakota   Watertown  Petrol   
7905         950000      25000          South Dakota   Watertown  Diesel   

     seller_type transmission        owner     mileage   engine    max_power  \
0     Individual       Manual  First Owner   23.4 kmpl  1248 CC       74 bhp   
1     Individual       Manual  First Owner   17.0 kmpl  1591 CC    121.3 bhp   
2     Individual       Manual  Third Owner  18.16 kmpl  1196 CC     86.8 bhp   
3     Individual       Manual  First Owner   22.0 kmpl  1498 CC  108.495 bhp   
4     Individual       Manual  First Owner   25.1 kmpl  1498 CC     98.6 bhp   
...          ...          ...          ...         ...      ...          ...   
7901  Individual       Manual  First Owner  28.09 kmpl  1248 CC     88.5 bhp   
7902  Individual       Manual  First Owner   24.3 kmpl  1248 CC     88.5 bhp   
7903  Individual       Manual  First Owner  22.74 kmpl   796 CC     47.3 bhp   
7904  Individual       Manual  First Owner   18.9 kmpl  1197 CC       82 bhp   
7905  Individual       Manual  First Owner   24.3 kmpl  1248 CC     88.5 bhp   

      seats sold   Region  
0         5    Y     East  
1         5    Y     East  
2         5    Y     East  
3         5    Y     East  
4         5    Y     East  
...     ...  ...      ...  
7901      5    N  Central  
7902      5    N  Central  
7903      5    N  Central  
7904      5    N  Central  
7905      5    N  Central  

[7906 rows x 17 columns]

In [16]:
car_df=pd.merge(df2, df3, on='State or Province', how='inner')

FINAL TABLE AFTER COMBINING ALL 3 TABLES

In [17]:
car_df

Sales_ID                                       name  year  \
0            1                     Maruti Swift Dzire VDI  2014   
1           38           Hyundai Verna VTVT 1.6 SX Option  2019   
2           66                 Ford Figo 1.2P Titanium MT  2017   
3           68  Volkswagen Ameo 1.5 TDI Highline 16 Alloy  2016   
4           69                       Honda City i DTEC VX  2015   
...        ...                                        ...   ...   
7901      4547                       Maruti Ciaz VDI SHVS  2016   
7902      4548                   Maruti Vitara Brezza ZDi  2017   
7903      4549                        Maruti Alto 800 LXI  2016   
7904      6888                    Hyundai Grand i10 Magna  2016   
7905      6889    Maruti Vitara Brezza ZDi Plus Dual Tone  2018   

      selling_price  km_driven     State or Province        City    fuel  \
0            450000     145500  District of Columbia  Washington  Diesel   
1           1149000       5000  District of Columbia  Washington  Petrol   
2            600000      25000  District of Columbia  Washington  Petrol   
3            540000      40000  District of Columbia  Washington  Diesel   
4            630000     135000  District of Columbia  Washington  Diesel   
...             ...        ...                   ...         ...     ...   
7901         620000      90000          South Dakota   Brookings  Diesel   
7902         750000      90000          South Dakota   Brookings  Diesel   
7903         270000      35000          South Dakota   Brookings  Petrol   
7904         470000      10000          South Dakota   Watertown  Petrol   
7905         950000      25000          South Dakota   Watertown  Diesel   

     seller_type transmission        owner     mileage   engine    max_power  \
0     Individual       Manual  First Owner   23.4 kmpl  1248 CC       74 bhp   
1     Individual       Manual  First Owner   17.0 kmpl  1591 CC    121.3 bhp   
2     Individual       Manual  Third Owner  18.16 kmpl  1196 CC     86.8 bhp   
3     Individual       Manual  First Owner   22.0 kmpl  1498 CC  108.495 bhp   
4     Individual       Manual  First Owner   25.1 kmpl  1498 CC     98.6 bhp   
...          ...          ...          ...         ...      ...          ...   
7901  Individual       Manual  First Owner  28.09 kmpl  1248 CC     88.5 bhp   
7902  Individual       Manual  First Owner   24.3 kmpl  1248 CC     88.5 bhp   
7903  Individual       Manual  First Owner  22.74 kmpl   796 CC     47.3 bhp   
7904  Individual       Manual  First Owner   18.9 kmpl  1197 CC       82 bhp   
7905  Individual       Manual  First Owner   24.3 kmpl  1248 CC     88.5 bhp   

      seats sold   Region  
0         5    Y     East  
1         5    Y     East  
2         5    Y     East  
3         5    Y     East  
4         5    Y     East  
...     ...  ...      ...  
7901      5    N  Central  
7902      5    N  Central  
7903      5    N  Central  
7904      5    N  Central  
7905      5    N  Central  

[7906 rows x 17 columns]

In [18]:
car_df.describe()

Sales_ID         year  selling_price     km_driven        seats
count  7906.000000  7906.000000   7.906000e+03  7.906000e+03  7906.000000
mean   4070.106248  2013.983936   6.498137e+05  6.918866e+04     5.416393
std    2345.770159     3.863695   8.135827e+05  5.679230e+04     0.959208
min       1.000000  1994.000000   2.999900e+04  1.000000e+00     2.000000
25%    2037.250000  2012.000000   2.700000e+05  3.500000e+04     5.000000
50%    4076.500000  2015.000000   4.500000e+05  6.000000e+04     5.000000
75%    6102.000000  2017.000000   6.900000e+05  9.542500e+04     5.000000
max    8128.000000  2020.000000   1.000000e+07  2.360457e+06    14.000000

# A)Significant Features

In [19]:
car_df.corr()

Sales_ID      year  selling_price  km_driven     seats
Sales_ID       1.000000 -0.001304      -0.012125   0.009556 -0.014154
year          -0.001304  1.000000       0.412302  -0.428548 -0.007923
selling_price -0.012125  0.412302       1.000000  -0.222158  0.041617
km_driven      0.009556 -0.428548      -0.222158   1.000000  0.227259
seats         -0.014154 -0.007923       0.041617   0.227259  1.000000

In [20]:
print(car_df['name'].unique().tolist())

['Maruti Swift Dzire VDI', 'Hyundai Verna VTVT 1.6 SX Option', 'Ford Figo 1.2P Titanium MT', 'Volkswagen Ameo 1.5 TDI Highline 16 Alloy', 'Honda City i DTEC VX', 'Maruti Wagon R AMT VXI', 'Maruti Wagon R VXI', 'Tata Manza Aura (ABS) Quadrajet', 'Maruti Swift VXI', 'Nissan Terrano XL 85 PS', 'Maruti Ertiga VDI', 'Honda Amaze S i-VTEC', 'Toyota Corolla Altis 1.8 G CVT', 'Volvo XC40 D4 R-Design', 'BMW 5 Series 523i', 'Hyundai Elantra S', 'Volkswagen Vento Petrol Highline', 'BMW 5 Series 520d Luxury Line', 'Maruti 800 Std', 'Maruti Ritz VDi', 'Renault KWID RXT', 'Maruti Eeco 5 Seater AC BSIV', 'Hyundai Santro Xing GLS', 'Honda Brio Exclusive Edition', 'Hyundai Grand i10 AT Asta', 'Mahindra Bolero DI DX 7 Seater', 'Renault KWID RXL', 'Toyota Innova 2.5 GX (Diesel) 8 Seater', 'Hyundai Creta 1.6 CRDi AT SX Plus', 'Maruti Swift VXI BSIV', 'Renault Duster 85PS Diesel RxL', 'Mahindra Verito 1.5 D2 BSIII', 'Maruti 800 AC BSIII', 'Hyundai Verna 1.6 CRDi S', 'Maruti Ertiga SHVS VDI', 'Maruti Swift 

In [21]:
len(car_df['name'].unique())

1982

In [22]:
car_df['seller_type'].unique()

array(['Individual', 'Dealer', 'Trustmark Dealer'], dtype=object)

In [23]:
car_df['owner'].unique()

array(['First Owner', 'Third Owner', 'Second Owner',
       'Fourth & Above Owner', 'Test Drive Car'], dtype=object)

In [24]:
car_df['transmission'].unique()

array(['Manual', 'Automatic'], dtype=object)

In [25]:
car_df['Region'].unique()

array(['East', 'Central', 'West', 'South'], dtype=object)

In [26]:
car_df['City'].unique()

array(['Washington', 'New York City', 'Albany', ..., 'Rapid City',
       'Sioux Falls', 'Brookings'], dtype=object)

In [27]:
len(car_df['City'].unique())

1310

In [28]:
car_df['mileage'].unique()

array(['23.4 kmpl', '17.0 kmpl', '18.16 kmpl', '22.0 kmpl', '25.1 kmpl',
       '20.51 kmpl', '21.79 kmpl', '18.8 kmpl', '21.21 kmpl',
       '20.45 kmpl', '20.77 kmpl', '17.8 kmpl', '14.28 kmpl', '18.0 kmpl',
       '10.8 kmpl', '16.3 kmpl', '15.04 kmpl', '18.12 kmpl', '16.1 kmpl',
       '21.1 kmpl', '25.17 kmpl', '15.37 kmpl', '17.92 kmpl', '19.4 kmpl',
       '18.9 kmpl', '13.6 kmpl', '12.99 kmpl', '17.01 kmpl', '21.0 kmpl',
       '23.9 kmpl', '24.52 kmpl', '25.2 kmpl', '18.5 kmpl', '23.84 kmpl',
       '18.2 kmpl', '19.7 kmpl', '20.3 kmpl', '14.3 kmpl', '16.8 kmpl',
       '20.36 kmpl', '20.4 kmpl', '16.02 kmpl', '19.09 kmpl', '17.7 kmpl',
       '15.1 kmpl', '20.62 kmpl', '22.54 kmpl', '22.07 kmpl',
       '13.58 kmpl', '15.4 kmpl', '26.1 kmpl', '22.7 kmpl', '23.1 kmpl',
       '21.14 kmpl', '20.14 kmpl', '19.01 kmpl', '25.5 kmpl', '21.5 kmpl',
       '16.0 kmpl', '23.95 kmpl', '14.8 kmpl', '24.7 kmpl', '20.65 kmpl',
       '21.4 kmpl', '22.32 kmpl', '21.9 kmpl', '14.53 kmpl', '

In [29]:
car_df['seats'].unique()

array([ 5,  7,  4,  8,  9,  6, 10,  2, 14], dtype=int64)

In [30]:
car_df['max_power'].unique()

array(['74 bhp', '121.3 bhp', '86.8 bhp', '108.495 bhp', '98.6 bhp',
       '67 bhp', '67.05 bhp', '90 bhp', '81.80 bhp', '83.8 bhp',
       '88.76 bhp', '86.7 bhp', '138.03 bhp', '190 bhp', '150 bhp',
       '147.5 bhp', '103.6 bhp', '37 bhp', '73.9 bhp', '53.3 bhp',
       '73 bhp', '62.1 bhp', '82 bhp', '63 bhp', '100.6 bhp', '126.2 bhp',
       '64.1 bhp', '88.5 bhp', '84 bhp', '82.4 bhp', '46.3 bhp', '68 bhp',
       '103.5 bhp', '78.9 bhp', '67.04 bhp', '55.2 bhp', '93.7 bhp',
       '69 bhp', '78 bhp', '81.86 bhp', '82.85 bhp', '140 bhp',
       '88.73 bhp', '72.4 bhp', '120 bhp', '99.23 bhp', '89.84 bhp',
       '103.52 bhp', '88.2 bhp', '108.45 bhp', '88.8 bhp', '83.81 bhp',
       '34.2 bhp', '160.77 bhp', '117.3 bhp', '116.3 bhp', '83.14 bhp',
       '100 bhp', '120.7 bhp', '98.63 bhp', '103.25 bhp', '83.1 bhp',
       '47.3 bhp', '155 bhp', '69.01 bhp', '126.32 bhp', '138.1 bhp',
       '98.96 bhp', '177 bhp', '108.5 bhp', '157.7 bhp', '164.7 bhp',
       '400 bhp', '75 bhp

In [31]:
car_df.isnull().sum()

Sales_ID             0
name                 0
year                 0
selling_price        0
km_driven            0
State or Province    0
City                 0
fuel                 0
seller_type          0
transmission         0
owner                0
mileage              0
engine               0
max_power            0
seats                0
sold                 0
Region               0
dtype: int64

# B)Describing the variable

In [32]:
car_df.describe()

Sales_ID         year  selling_price     km_driven        seats
count  7906.000000  7906.000000   7.906000e+03  7.906000e+03  7906.000000
mean   4070.106248  2013.983936   6.498137e+05  6.918866e+04     5.416393
std    2345.770159     3.863695   8.135827e+05  5.679230e+04     0.959208
min       1.000000  1994.000000   2.999900e+04  1.000000e+00     2.000000
25%    2037.250000  2012.000000   2.700000e+05  3.500000e+04     5.000000
50%    4076.500000  2015.000000   4.500000e+05  6.000000e+04     5.000000
75%    6102.000000  2017.000000   6.900000e+05  9.542500e+04     5.000000
max    8128.000000  2020.000000   1.000000e+07  2.360457e+06    14.000000

In [33]:
car_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7906 entries, 0 to 7905
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Sales_ID           7906 non-null   int64 
 1   name               7906 non-null   object
 2   year               7906 non-null   int64 
 3   selling_price      7906 non-null   int64 
 4   km_driven          7906 non-null   int64 
 5   State or Province  7906 non-null   object
 6   City               7906 non-null   object
 7   fuel               7906 non-null   object
 8   seller_type        7906 non-null   object
 9   transmission       7906 non-null   object
 10  owner              7906 non-null   object
 11  mileage            7906 non-null   object
 12  engine             7906 non-null   object
 13  max_power          7906 non-null   object
 14  seats              7906 non-null   int64 
 15  sold               7906 non-null   object
 16  Region             7906 non-null   object


# Finding which type of fuel cars are selling most

In [34]:
car_df.groupby('fuel')['sold'].value_counts()

fuel    sold
CNG     N         45
        Y          7
Diesel  N       3247
        Y       1052
LPG     N         29
        Y          6
Petrol  N       2585
        Y        935
Name: sold, dtype: int64

In [35]:
car_df.query('sold=="Y"').groupby('fuel')['sold'].value_counts()

fuel    sold
CNG     Y          7
Diesel  Y       1052
LPG     Y          6
Petrol  Y        935
Name: sold, dtype: int64

In [36]:
car_df['sale']=(car_df['sold']=='Y').map({True:1,False:0})

In [37]:
car_df.head()

Sales_ID                                       name  year  selling_price  \
0         1                     Maruti Swift Dzire VDI  2014         450000   
1        38           Hyundai Verna VTVT 1.6 SX Option  2019        1149000   
2        66                 Ford Figo 1.2P Titanium MT  2017         600000   
3        68  Volkswagen Ameo 1.5 TDI Highline 16 Alloy  2016         540000   
4        69                       Honda City i DTEC VX  2015         630000   

   km_driven     State or Province        City    fuel seller_type  \
0     145500  District of Columbia  Washington  Diesel  Individual   
1       5000  District of Columbia  Washington  Petrol  Individual   
2      25000  District of Columbia  Washington  Petrol  Individual   
3      40000  District of Columbia  Washington  Diesel  Individual   
4     135000  District of Columbia  Washington  Diesel  Individual   

  transmission        owner     mileage   engine    max_power  seats sold  \
0       Manual  First Owner   23.4 kmpl  1248 CC       74 bhp      5    Y   
1       Manual  First Owner   17.0 kmpl  1591 CC    121.3 bhp      5    Y   
2       Manual  Third Owner  18.16 kmpl  1196 CC     86.8 bhp      5    Y   
3       Manual  First Owner   22.0 kmpl  1498 CC  108.495 bhp      5    Y   
4       Manual  First Owner   25.1 kmpl  1498 CC     98.6 bhp      5    Y   

  Region  sale  
0   East     1  
1   East     1  
2   East     1  
3   East     1  
4   East     1

In [38]:
f=car_df.query('sale==1').groupby('fuel').agg({'sale':np.sum,'selling_price':np.mean,'km_driven':np.mean,})

In [39]:
print(f)

        sale  selling_price     km_driven
fuel                                     
CNG        7  362142.714286  52142.857143
Diesel  1052  836903.476236  83923.622624
LPG        6  201166.666667  96500.000000
Petrol   935  501153.604278  51532.865241


In [40]:
f['percent']=100*f['sale']/f['sale'].sum()

In [41]:
f

sale  selling_price     km_driven  percent
fuel                                              
CNG        7  362142.714286  52142.857143     0.35
Diesel  1052  836903.476236  83923.622624    52.60
LPG        6  201166.666667  96500.000000     0.30
Petrol   935  501153.604278  51532.865241    46.75

In [42]:
f.sort_values('percent',ascending=False)

sale  selling_price     km_driven  percent
fuel                                              
Diesel  1052  836903.476236  83923.622624    52.60
Petrol   935  501153.604278  51532.865241    46.75
CNG        7  362142.714286  52142.857143     0.35
LPG        6  201166.666667  96500.000000     0.30

It is clear that more percent of Diesel cars were selling compared to other fuel types

# Finding which type of Engine capacity cars were selling most

In [43]:
len(car_df['engine'].unique())

121

In [44]:
car_df.groupby('engine')['sold'].value_counts()

engine   sold
1047 CC  N        11
         Y         7
1061 CC  N        45
         Y        14
1086 CC  N       105
                ... 
995 CC   Y         7
998 CC   N       337
         Y       116
999 CC   N        60
         Y        17
Name: sold, Length: 218, dtype: int64

In [45]:
e=car_df.query('sale==1').groupby('engine').agg({'sale':np.sum,'selling_price':np.mean,'km_driven':np.mean})

In [46]:
e

sale  selling_price     km_driven
engine                                    
1047 CC     7  469285.714286  75439.857143
1061 CC    14  151142.785714  90428.571429
1086 CC    33  240302.909091  63469.696970
1120 CC    13  430384.538462  83653.846154
1150 CC     2  102500.000000  65000.000000
...       ...            ...           ...
936 CC      9  218333.333333  71591.000000
993 CC      8   89625.000000  95125.000000
995 CC      7  127000.000000  78714.285714
998 CC    116  279543.060345  58011.215517
999 CC     17  405705.823529  30072.823529

[100 rows x 3 columns]

In [47]:
e['percent']=100*e['sale']/e['sale'].sum()

In [48]:
e

sale  selling_price     km_driven  percent
engine                                             
1047 CC     7  469285.714286  75439.857143     0.35
1061 CC    14  151142.785714  90428.571429     0.70
1086 CC    33  240302.909091  63469.696970     1.65
1120 CC    13  430384.538462  83653.846154     0.65
1150 CC     2  102500.000000  65000.000000     0.10
...       ...            ...           ...      ...
936 CC      9  218333.333333  71591.000000     0.45
993 CC      8   89625.000000  95125.000000     0.40
995 CC      7  127000.000000  78714.285714     0.35
998 CC    116  279543.060345  58011.215517     5.80
999 CC     17  405705.823529  30072.823529     0.85

[100 rows x 4 columns]

In [49]:
e.sort_values('percent',ascending=False)

sale  selling_price      km_driven  percent
engine                                              
1248 CC   239  523271.928870   80591.133891    11.95
1197 CC   214  505345.752336   42694.261682    10.70
998 CC    116  279543.060345   58011.215517     5.80
796 CC    115  162334.800000   61003.452174     5.75
2179 CC    94  678606.361702  109553.159574     4.70
...       ...            ...            ...      ...
2497 CC     1  975000.000000   72200.000000     0.05
2496 CC     1  600000.000000   80000.000000     0.05
1595 CC     1  225000.000000   80000.000000     0.05
2362 CC     1  450000.000000   71000.000000     0.05
2596 CC     1  910000.000000   20000.000000     0.05

[100 rows x 4 columns]

1248cc engine encrypted cars were selling most with almost 12% sales compared to other engine types

# For what kind of Average Mileage People were intrested to buy Cars

In [50]:
len(car_df['mileage'].unique())

393

In [51]:
car_df.groupby('mileage')['sold'].value_counts()

mileage    sold
0.0 kmpl   N       13
           Y        4
10.0 kmpl  N        2
10.1 kmpl  N        1
           Y        1
                   ..
42.0 kmpl  Y        1
9.0 kmpl   N        3
           Y        1
9.5 kmpl   N        3
           Y        3
Name: sold, Length: 659, dtype: int64

In [52]:
m=car_df.query('sale==1').groupby('mileage').agg({'sale':np.sum,'selling_price':np.mean,'km_driven':np.mean})

In [53]:
m

sale  selling_price  km_driven
mileage                                    
0.0 kmpl        4   5.150000e+05    89225.0
10.1 kmpl       1   3.250000e+05    70000.0
10.5 kmpl       7   3.279997e+05   104000.0
10.8 kmpl       1   9.750000e+05    72200.0
10.91 kmpl      3   2.926667e+06    48000.0
...           ...            ...        ...
30.46 km/kg     1   1.650000e+05    60000.0
33.44 km/kg     1   3.300000e+05    10000.0
42.0 kmpl       1   1.000000e+07    30000.0
9.0 kmpl        1   2.750000e+05   100000.0
9.5 kmpl        3   4.100000e+06    17000.0

[282 rows x 3 columns]

In [54]:
m['percent']=100*m['sale']/m['sale'].sum()

In [55]:
m

sale  selling_price  km_driven  percent
mileage                                             
0.0 kmpl        4   5.150000e+05    89225.0     0.20
10.1 kmpl       1   3.250000e+05    70000.0     0.05
10.5 kmpl       7   3.279997e+05   104000.0     0.35
10.8 kmpl       1   9.750000e+05    72200.0     0.05
10.91 kmpl      3   2.926667e+06    48000.0     0.15
...           ...            ...        ...      ...
30.46 km/kg     1   1.650000e+05    60000.0     0.05
33.44 km/kg     1   3.300000e+05    10000.0     0.05
42.0 kmpl       1   1.000000e+07    30000.0     0.05
9.0 kmpl        1   2.750000e+05   100000.0     0.05
9.5 kmpl        3   4.100000e+06    17000.0     0.15

[282 rows x 4 columns]

In [56]:
m.sort_values('percent',ascending=False)

sale  selling_price     km_driven  percent
mileage                                                
18.9 kmpl      51  315999.941176  68151.549020     2.55
19.7 kmpl      44  126295.454545  80886.727273     2.20
18.6 kmpl      41  520902.439024  34756.487805     2.05
21.1 kmpl      34  240500.000000  60344.117647     1.70
15.96 kmpl     33  509909.030303  97332.575758     1.65
...           ...            ...           ...      ...
15.85 kmpl      1  670000.000000  50478.000000     0.05
15.9 kmpl       1  640000.000000   6001.000000     0.05
16.34 kmpl      1  800000.000000  70000.000000     0.05
20.88 km/kg     1  409999.000000  35000.000000     0.05
21.7 kmpl       1  475000.000000  48000.000000     0.05

[282 rows x 4 columns]

For the average mileage of 18.9 kmpl most cars were sold.

# In Which States the Sales were high

In [57]:
car_df.columns

Index(['Sales_ID', 'name', 'year', 'selling_price', 'km_driven',
       'State or Province', 'City', 'fuel', 'seller_type', 'transmission',
       'owner', 'mileage', 'engine', 'max_power', 'seats', 'sold', 'Region',
       'sale'],
      dtype='object')

In [58]:
car_df['State or Province'].unique()

array(['District of Columbia', 'New York', 'Illinois', 'California',
       'Texas', 'Massachusetts', 'Washington', 'Georgia',
       'North Carolina', 'Florida', 'Pennsylvania', 'Michigan', 'Iowa',
       'Oregon', 'New Mexico', 'Maryland', 'Montana', 'Arkansas',
       'Arizona', 'Maine', 'Vermont', 'New Jersey', 'Connecticut',
       'Oklahoma', 'Tennessee', 'Alabama', 'Wisconsin', 'South Carolina',
       'New Hampshire', 'Idaho', 'Minnesota', 'Ohio', 'Kansas', 'Utah',
       'Virginia', 'Nebraska', 'West Virginia', 'Colorado', 'Indiana',
       'Kentucky', 'Missouri', 'North Dakota', 'Rhode Island',
       'Louisiana', 'Wyoming', 'Nevada', 'Mississippi', 'Delaware',
       'South Dakota'], dtype=object)

In [59]:
len(car_df['State or Province'].unique())

49

In [60]:
car_df.groupby('State or Province')['sold'].value_counts()

State or Province  sold
Alabama            N        95
                   Y        11
Arizona            N       103
                   Y        12
Arkansas           N        80
                          ... 
West Virginia      N        29
                   Y         4
Wisconsin          N        92
                   Y        25
Wyoming            N        12
Name: sold, Length: 89, dtype: int64

In [61]:
s=car_df.query('sale==1').groupby('State or Province').agg({'sale':np.sum,'selling_price':np.mean,'km_driven':np.mean})

In [62]:
s

sale  selling_price      km_driven
State or Province                                       
Alabama                 11   4.782726e+05   88181.818182
Arizona                 12   6.933333e+05   71081.833333
Arkansas                27   1.141815e+06   55364.111111
California             274   6.286788e+05   70237.062044
Colorado                 8   6.453750e+05   32012.500000
Connecticut              4   3.550000e+05   53750.000000
District of Columbia    68   5.952793e+05   66486.676471
Florida                118   6.984958e+05   76132.144068
Georgia                 62   7.144515e+05   58323.322581
Idaho                   15   1.279600e+06   31745.333333
Illinois               148   5.778243e+05   83738.520270
Indiana                 45   6.303111e+05   70290.577778
Iowa                    30   8.924999e+05   61038.533333
Kansas                  26   4.450827e+05   56869.923077
Kentucky                12   7.504167e+05   86125.000000
Maine                   14   4.567857e+05   65214.285714
Maryland                28   1.266036e+06   53165.321429
Massachusetts           76   6.625921e+05   67858.355263
Michigan                71   6.663662e+05   63133.943662
Minnesota               25   6.464000e+05   70288.240000
Missouri                 3   3.966667e+05   39297.333333
Montana                  5   1.629800e+06   24369.800000
Nebraska                 2   4.875000e+05   53247.000000
New Hampshire            4   2.342500e+05   97500.000000
New Jersey              15   3.785333e+05   65600.000000
New Mexico               3   1.134667e+06   33553.000000
New York               293   6.958071e+05   69979.000000
North Carolina          75   1.158547e+06   52296.853333
Ohio                    40   7.684250e+05   59654.850000
Oklahoma                27   4.428518e+05   64918.518519
Oregon                  30   8.879333e+05   70943.233333
Pennsylvania            80   4.533500e+05   68760.437500
South Carolina           4   5.962500e+05  104750.000000
Tennessee               29   4.574481e+05   86120.689655
Texas                  138   6.576449e+05   73880.847826
Utah                    20   5.930500e+05   55315.600000
Vermont                  4   5.512500e+05   44125.000000
Virginia                 5   4.709998e+05   58335.600000
Washington             120   6.410083e+05   67080.391667
West Virginia            4   2.800000e+05  121500.000000
Wisconsin               25   3.452000e+05   73941.720000

In [63]:
s['percent']=100*s['sale']/s['sale'].sum()

In [64]:
s

sale  selling_price      km_driven  percent
State or Province                                                
Alabama                 11   4.782726e+05   88181.818182     0.55
Arizona                 12   6.933333e+05   71081.833333     0.60
Arkansas                27   1.141815e+06   55364.111111     1.35
California             274   6.286788e+05   70237.062044    13.70
Colorado                 8   6.453750e+05   32012.500000     0.40
Connecticut              4   3.550000e+05   53750.000000     0.20
District of Columbia    68   5.952793e+05   66486.676471     3.40
Florida                118   6.984958e+05   76132.144068     5.90
Georgia                 62   7.144515e+05   58323.322581     3.10
Idaho                   15   1.279600e+06   31745.333333     0.75
Illinois               148   5.778243e+05   83738.520270     7.40
Indiana                 45   6.303111e+05   70290.577778     2.25
Iowa                    30   8.924999e+05   61038.533333     1.50
Kansas                  26   4.450827e+05   56869.923077     1.30
Kentucky                12   7.504167e+05   86125.000000     0.60
Maine                   14   4.567857e+05   65214.285714     0.70
Maryland                28   1.266036e+06   53165.321429     1.40
Massachusetts           76   6.625921e+05   67858.355263     3.80
Michigan                71   6.663662e+05   63133.943662     3.55
Minnesota               25   6.464000e+05   70288.240000     1.25
Missouri                 3   3.966667e+05   39297.333333     0.15
Montana                  5   1.629800e+06   24369.800000     0.25
Nebraska                 2   4.875000e+05   53247.000000     0.10
New Hampshire            4   2.342500e+05   97500.000000     0.20
New Jersey              15   3.785333e+05   65600.000000     0.75
New Mexico               3   1.134667e+06   33553.000000     0.15
New York               293   6.958071e+05   69979.000000    14.65
North Carolina          75   1.158547e+06   52296.853333     3.75
Ohio                    40   7.684250e+05   59654.850000     2.00
Oklahoma                27   4.428518e+05   64918.518519     1.35
Oregon                  30   8.879333e+05   70943.233333     1.50
Pennsylvania            80   4.533500e+05   68760.437500     4.00
South Carolina           4   5.962500e+05  104750.000000     0.20
Tennessee               29   4.574481e+05   86120.689655     1.45
Texas                  138   6.576449e+05   73880.847826     6.90
Utah                    20   5.930500e+05   55315.600000     1.00
Vermont                  4   5.512500e+05   44125.000000     0.20
Virginia                 5   4.709998e+05   58335.600000     0.25
Washington             120   6.410083e+05   67080.391667     6.00
West Virginia            4   2.800000e+05  121500.000000     0.20
Wisconsin               25   3.452000e+05   73941.720000     1.25

In [65]:
s.sort_values('percent',ascending=False)

sale  selling_price      km_driven  percent
State or Province                                                
New York               293   6.958071e+05   69979.000000    14.65
California             274   6.286788e+05   70237.062044    13.70
Illinois               148   5.778243e+05   83738.520270     7.40
Texas                  138   6.576449e+05   73880.847826     6.90
Washington             120   6.410083e+05   67080.391667     6.00
Florida                118   6.984958e+05   76132.144068     5.90
Pennsylvania            80   4.533500e+05   68760.437500     4.00
Massachusetts           76   6.625921e+05   67858.355263     3.80
North Carolina          75   1.158547e+06   52296.853333     3.75
Michigan                71   6.663662e+05   63133.943662     3.55
District of Columbia    68   5.952793e+05   66486.676471     3.40
Georgia                 62   7.144515e+05   58323.322581     3.10
Indiana                 45   6.303111e+05   70290.577778     2.25
Ohio                    40   7.684250e+05   59654.850000     2.00
Oregon                  30   8.879333e+05   70943.233333     1.50
Iowa                    30   8.924999e+05   61038.533333     1.50
Tennessee               29   4.574481e+05   86120.689655     1.45
Maryland                28   1.266036e+06   53165.321429     1.40
Oklahoma                27   4.428518e+05   64918.518519     1.35
Arkansas                27   1.141815e+06   55364.111111     1.35
Kansas                  26   4.450827e+05   56869.923077     1.30
Wisconsin               25   3.452000e+05   73941.720000     1.25
Minnesota               25   6.464000e+05   70288.240000     1.25
Utah                    20   5.930500e+05   55315.600000     1.00
New Jersey              15   3.785333e+05   65600.000000     0.75
Idaho                   15   1.279600e+06   31745.333333     0.75
Maine                   14   4.567857e+05   65214.285714     0.70
Arizona                 12   6.933333e+05   71081.833333     0.60
Kentucky                12   7.504167e+05   86125.000000     0.60
Alabama                 11   4.782726e+05   88181.818182     0.55
Colorado                 8   6.453750e+05   32012.500000     0.40
Montana                  5   1.629800e+06   24369.800000     0.25
Virginia                 5   4.709998e+05   58335.600000     0.25
New Hampshire            4   2.342500e+05   97500.000000     0.20
South Carolina           4   5.962500e+05  104750.000000     0.20
Connecticut              4   3.550000e+05   53750.000000     0.20
Vermont                  4   5.512500e+05   44125.000000     0.20
West Virginia            4   2.800000e+05  121500.000000     0.20
New Mexico               3   1.134667e+06   33553.000000     0.15
Missouri                 3   3.966667e+05   39297.333333     0.15
Nebraska                 2   4.875000e+05   53247.000000     0.10

In New York State there are more number of sales with 14.65%

# D)what is the city with most number of sales

In [66]:
car_df['City'].unique()

array(['Washington', 'New York City', 'Albany', ..., 'Rapid City',
       'Sioux Falls', 'Brookings'], dtype=object)

In [67]:
len(car_df['City'].unique())

1310

In [68]:
car_df.groupby('City')['sold'].value_counts()

City        sold
Abilene     N       7
Abington    N       1
Acton       N       1
Addison     N       1
Adrian      N       5
                   ..
York        N       5
Youngstown  N       3
Yuba City   N       2
Yucaipa     N       2
Yuma        N       5
Name: sold, Length: 1463, dtype: int64

In [69]:
c=car_df.query('sale==1').groupby('City').agg({'sale':np.sum,'selling_price':np.mean,'km_driven':np.mean})

In [70]:
c

sale  selling_price      km_driven
City                                            
Albany           5   6.678000e+05   42566.600000
Albemarle        3   4.875000e+06   20100.000000
Allen Park       1   3.000000e+05  100000.000000
Aloha            6   1.610833e+06   41391.000000
Altoona          5   7.820000e+05   47360.000000
...            ...            ...            ...
White Plains     5   3.140000e+05   64940.000000
Willingboro      1   9.000000e+04   40000.000000
Wilmette         7   6.207143e+05   61857.142857
Woodridge        6   8.108333e+05   94500.000000
Yonkers          7   2.728571e+05   86000.000000

[275 rows x 3 columns]

In [71]:
c['percent']=100*c['sale']/c['sale'].sum()

In [72]:
c

sale  selling_price      km_driven  percent
City                                                     
Albany           5   6.678000e+05   42566.600000     0.25
Albemarle        3   4.875000e+06   20100.000000     0.15
Allen Park       1   3.000000e+05  100000.000000     0.05
Aloha            6   1.610833e+06   41391.000000     0.30
Altoona          5   7.820000e+05   47360.000000     0.25
...            ...            ...            ...      ...
White Plains     5   3.140000e+05   64940.000000     0.25
Willingboro      1   9.000000e+04   40000.000000     0.05
Wilmette         7   6.207143e+05   61857.142857     0.35
Woodridge        6   8.108333e+05   94500.000000     0.30
Yonkers          7   2.728571e+05   86000.000000     0.35

[275 rows x 4 columns]

In [73]:
c.sort_values('percent',ascending=False)

sale  selling_price      km_driven  percent
City                                                      
New York City   196  688987.224490   68372.102041     9.80
Los Angeles     191  654246.026178   68726.460733     9.55
Seattle          86  627348.802326   66981.011628     4.30
Chicago          83  618168.590361   72614.216867     4.15
Boston           69  696695.608696   64191.811594     3.45
...             ...            ...            ...      ...
Brookfield        1  190000.000000  120000.000000     0.05
Lynn              1  480000.000000   40000.000000     0.05
Iowa City         1  650000.000000  110000.000000     0.05
Glassboro         1  580000.000000   30000.000000     0.05
Lake Jackson      1  599000.000000   60000.000000     0.05

[275 rows x 4 columns]

In New York City there are more number of sales

# C)Which type of cars are selling more

In [74]:
len(car_df['name'].unique())

1982

In [75]:
car_df.groupby('name')['sold'].value_counts()

name                               sold
Ambassador CLASSIC 1500 DSL AC     N       1
Ambassador Classic 2000 DSZ AC PS  N       1
Ambassador Grand 1500 DSZ BSIII    N       1
Ambassador Grand 2000 DSZ PW CL    N       1
Ashok Leyland Stile LE             N       1
                                          ..
Volvo XC40 D4 Inscription BSIV     Y       9
Volvo XC40 D4 R-Design             N       1
                                   Y       1
Volvo XC60 Inscription D5 BSIV     N       1
Volvo XC90 T8 Excellence BSIV      Y       1
Name: sold, Length: 2732, dtype: int64

In [76]:
n=car_df.query('sale==1').groupby('name').agg({'sale':np.sum,'selling_price':np.mean,'km_driven':np.mean})

In [77]:
n

sale  selling_price  km_driven
name                                                                
Audi A4 2.0 TDI                          1      1600000.0    44000.0
Audi A6 2.0 TDI  Design Edition          1      1689999.0    50000.0
Audi A6 2.0 TDI Technology               1      1750000.0    33900.0
Audi Q3 2.0 TDI Quattro Premium Plus     1      2825000.0    22000.0
Audi Q5 2.0 TDI                          1      1300000.0   120000.0
...                                    ...            ...        ...
Volkswagen Vento Petrol Highline         2       455000.0    30500.0
Volvo V40 D3 R-Design                    8      2475000.0     2000.0
Volvo XC40 D4 Inscription BSIV           9      3800000.0    20000.0
Volvo XC40 D4 R-Design                   1      3400000.0    22000.0
Volvo XC90 T8 Excellence BSIV            1     10000000.0    30000.0

[968 rows x 3 columns]

In [78]:
n['percent']=100*n['sale']/n['sale'].sum()

In [79]:
n

sale  selling_price  km_driven  percent
name                                                                         
Audi A4 2.0 TDI                          1      1600000.0    44000.0     0.05
Audi A6 2.0 TDI  Design Edition          1      1689999.0    50000.0     0.05
Audi A6 2.0 TDI Technology               1      1750000.0    33900.0     0.05
Audi Q3 2.0 TDI Quattro Premium Plus     1      2825000.0    22000.0     0.05
Audi Q5 2.0 TDI                          1      1300000.0   120000.0     0.05
...                                    ...            ...        ...      ...
Volkswagen Vento Petrol Highline         2       455000.0    30500.0     0.10
Volvo V40 D3 R-Design                    8      2475000.0     2000.0     0.40
Volvo XC40 D4 Inscription BSIV           9      3800000.0    20000.0     0.45
Volvo XC40 D4 R-Design                   1      3400000.0    22000.0     0.05
Volvo XC90 T8 Excellence BSIV            1     10000000.0    30000.0     0.05

[968 rows x 4 columns]

In [80]:
n.sort_values('percent',ascending=False)

sale  selling_price  \
name                                                              
Maruti Swift Dzire VDI                        42   5.805952e+05   
Maruti Alto 800 LXI                           21   2.477142e+05   
BMW X4 M Sport X xDrive20d                    19   5.468421e+06   
Maruti Wagon R VXI BS IV                      16   3.493749e+05   
Maruti Wagon R LXI                            15   1.649333e+05   
...                                          ...            ...   
Mahindra Bolero DI AC BSIII                    1   2.500000e+05   
Mahindra Bolero DI Non AC BSIII                1   3.200000e+05   
Mahindra Bolero PLUS AC BSIII                  1   3.200000e+05   
Mahindra Bolero Power Plus Plus AC BSIV PS     1   8.000000e+05   
Volvo XC90 T8 Excellence BSIV                  1   1.000000e+07   

                                                km_driven  percent  
name                                                                
Maruti Swift Dzire VDI                       74238.095238     2.10  
Maruti Alto 800 LXI                          38351.476190     1.05  
BMW X4 M Sport X xDrive20d                    7973.684211     0.95  
Maruti Wagon R VXI BS IV                     38361.812500     0.80  
Maruti Wagon R LXI                           81249.600000     0.75  
...                                                   ...      ...  
Mahindra Bolero DI AC BSIII                  80000.000000     0.05  
Mahindra Bolero DI Non AC BSIII             300000.000000     0.05  
Mahindra Bolero PLUS AC BSIII               175000.000000     0.05  
Mahindra Bolero Power Plus Plus AC BSIV PS    2000.000000     0.05  
Volvo XC90 T8 Excellence BSIV                30000.000000     0.05  

[968 rows x 4 columns]

In [81]:
car_df['km_driven'].unique()

array([ 145500,    5000,   25000,   40000,  135000,   30000,   77000,
        100000,   10000,   70000,  110000,    1303,   39000,   22000,
         72200,   49900,   49000,   12000,   50000,  120000,   44000,
         80000,   13500,   75000,   33033,   63063,  207890,  160000,
         67000,   52000,  170000,  119000,   90000,   42000,   11000,
          2000,   98000,   60000,   32000,  155201,  101504,   45000,
         68000,   72000,   35000,   28900,   86300,   23300,   10300,
         99000,   27800,   49800,   64000,  140000,   38000,   37800,
        150000,   69000,   13000,   76139,    7500,    9000,   95000,
          8000,   41000,   71000,   20000,   99361,   58000,  162000,
         56494,  265000,   38817,   25538,   69779,   99500,  146000,
         85000,   15000,   43000,   76000,   31711,   66693,  221889,
        116104,  150546,   24000,   16500,   98500,   21000,  180000,
          6000,   51000,   86000,  107000,  168000,  144000,  107825,
          1620,  108

In [82]:
car_df.groupby('km_driven')['sold'].value_counts()

km_driven  sold
1          N       1
1000       N       5
           Y       1
1300       N       1
1303       Y       3
                  ..
475000     N       1
500000     N       2
577414     N       1
1500000    Y       1
2360457    N       1
Name: sold, Length: 1116, dtype: int64

In [83]:
k=car_df.query('sale==1').groupby('km_driven').agg({'sale':np.sum,'selling_price':np.mean})

In [84]:
k

sale  selling_price
km_driven                     
1000          1      1085000.0
1303          3       690000.0
1500          1       730000.0
1600          1      1900000.0
1620          1      1040000.0
...         ...            ...
300000        4       305000.0
334000        1       450000.0
360003        1       220000.0
375000        1       650000.0
1500000       1       500000.0

[384 rows x 2 columns]

In [85]:
k['percent']=100*k['sale']/k['sale'].sum()

In [86]:
k

sale  selling_price  percent
km_driven                              
1000          1      1085000.0     0.05
1303          3       690000.0     0.15
1500          1       730000.0     0.05
1600          1      1900000.0     0.05
1620          1      1040000.0     0.05
...         ...            ...      ...
300000        4       305000.0     0.20
334000        1       450000.0     0.05
360003        1       220000.0     0.05
375000        1       650000.0     0.05
1500000       1       500000.0     0.05

[384 rows x 3 columns]

In [87]:
k.sort_values('percent',ascending=False)

sale  selling_price  percent
km_driven                              
120000      127  379803.086614     6.35
50000       111  560081.018018     5.55
80000       111  368812.126126     5.55
70000       107  435280.336449     5.35
90000        94  358851.042553     4.70
...         ...            ...      ...
53619         1  850000.000000     0.05
54290         1  350000.000000     0.05
54470         1  950000.000000     0.05
54700         1  600000.000000     0.05
1500000       1  500000.000000     0.05

[384 rows x 3 columns]

# For which type of owner cars there are more number of sales

In [88]:
len(car_df['owner'].unique())

5

In [89]:
car_df.groupby('owner')['sold'].value_counts()

owner                 sold
First Owner           N       3889
                      Y       1326
Fourth & Above Owner  N        111
                      Y         49
Second Owner          N       1526
                      Y        490
Test Drive Car        N          5
Third Owner           N        375
                      Y        135
Name: sold, dtype: int64

In [90]:
o=car_df.query('sale==1').groupby('owner').agg({'sale':np.sum,'km_driven':np.mean,'selling_price':np.mean})

In [91]:
o

sale      km_driven  selling_price
owner                                                   
First Owner           1326   57276.268477  837219.423077
Fourth & Above Owner    49   99132.653061  197989.734694
Second Owner           490   86976.832653  390865.563265
Third Owner            135  103632.592593  306403.688889

In [92]:
o['percent']=100*o['sale']/o['sale'].sum()

In [93]:
o

sale      km_driven  selling_price  percent
owner                                                            
First Owner           1326   57276.268477  837219.423077    66.30
Fourth & Above Owner    49   99132.653061  197989.734694     2.45
Second Owner           490   86976.832653  390865.563265    24.50
Third Owner            135  103632.592593  306403.688889     6.75

In [94]:
o.sort_values('percent',ascending=False)

sale      km_driven  selling_price  percent
owner                                                            
First Owner           1326   57276.268477  837219.423077    66.30
Second Owner           490   86976.832653  390865.563265    24.50
Third Owner            135  103632.592593  306403.688889     6.75
Fourth & Above Owner    49   99132.653061  197989.734694     2.45

If the used car has single owner then there are more number of sales

e)Variables like city, km_driven, owner_type, fuel_type, mileage etc.., are most important features in deceiding sales and selling price but fuel_type is most important feature in deceiding selling price of a car with over 66% sales 

# Data Visuvalization Using Python

# A)Significant Values in Predicting Selling Price of Car

In [95]:
import pandas as pd
import numpy as np
from pandas import DataFrame
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [96]:
car_df.head()

Sales_ID                                       name  year  selling_price  \
0         1                     Maruti Swift Dzire VDI  2014         450000   
1        38           Hyundai Verna VTVT 1.6 SX Option  2019        1149000   
2        66                 Ford Figo 1.2P Titanium MT  2017         600000   
3        68  Volkswagen Ameo 1.5 TDI Highline 16 Alloy  2016         540000   
4        69                       Honda City i DTEC VX  2015         630000   

   km_driven     State or Province        City    fuel seller_type  \
0     145500  District of Columbia  Washington  Diesel  Individual   
1       5000  District of Columbia  Washington  Petrol  Individual   
2      25000  District of Columbia  Washington  Petrol  Individual   
3      40000  District of Columbia  Washington  Diesel  Individual   
4     135000  District of Columbia  Washington  Diesel  Individual   

  transmission        owner     mileage   engine    max_power  seats sold  \
0       Manual  First Owner   23.4 kmpl  1248 CC       74 bhp      5    Y   
1       Manual  First Owner   17.0 kmpl  1591 CC    121.3 bhp      5    Y   
2       Manual  Third Owner  18.16 kmpl  1196 CC     86.8 bhp      5    Y   
3       Manual  First Owner   22.0 kmpl  1498 CC  108.495 bhp      5    Y   
4       Manual  First Owner   25.1 kmpl  1498 CC     98.6 bhp      5    Y   

  Region  sale  
0   East     1  
1   East     1  
2   East     1  
3   East     1  
4   East     1